gr-dtv/lib/atsc/atsc_equalizer_impl.cc

```diff
@@ -27,42 +28,46 @@ atsc_equalizer::sptr atsc_equalizer::make()
     return gnuradio::make_block_sptr<atsc_equalizer_impl>();
 }
 
-static float bin_map(int bit) { return bit ? +5 : -5; }
+constexpr float bin_map(unsigned char bit) { return bit ? +5 : -5; }
 
-static void init_field_sync_common(float* p, int mask)
+template <size_t N>
+constexpr std::array<float, N> init_field_sync_common(bool invert)
```

Comment:
`std::array` does not have move assignment/constructor, thus every call here will cause `N*4` bytes to be copied (unless RVO). Is that the intention?

Reply:
`std::array` is stack (well… technically "auto storage duration") allocated (including the data), so cannot even in theory do a pointer swap for move. It can do per-element move of all its elements, but for floats that's not an optimization.

Your godbolt example shows two differences:
1. it likely triggers RVO, which always means no copy nor move. (I don't think it's [mandatory RVO](https://en.cppreference.com/w/cpp/language/copy_elision))
2. the function is inlined, so not really returned.

[Adding `-fno-inline` takes care of the second, and making RVO impossible](https://gcc.godbolt.org/z/6GzGcn9eo) (for example by calling a second time) causes a call to memcpy (or [in this example](https://gcc.godbolt.org/z/61n6x7rfe) a `rep movsq`).

So if the only correct way to call this function is in the mandatory RVO situations? That may be a bit of a foot gun for non-initialization at caller site. And I don't think as-is this is mandatory RVO since it matches the non-mandatory's section "In a return statement, when the operand is the name of a non-volatile object with automatic storage duration".

I guess since this is constexpr it may actually be mandatory RVO, but still only applies when caller makes it possible (in initialization).

As for `std::move`, despite its name it doesn't move. It just converts lvalue to rvalue, and you can copy an rvalue. But also yes technically `std::array` can be moved, but it's just an elementwise move, which for array of float means a de facto copy.

Also examples like [this](https://gcc.godbolt.org/z/fq8YovMed) show that `std::move()` doesn't force a move, only enables it.

But yes, technically I was wrong by saying `std::array` doesn't "have" a move assign/constructor. It does, it just doesn't do the expected thing, because it can't.

Reply:
Oh, wow! First of all: thanks for letting me occupy you so much! I'm thankful, and I'm learning a lot :)

So, the right approach here is instead letting `init_field_sync_common` return a `std::vector` value, making `training_sequence1,2` std::vectors and using `std::vector`s (actually moving) move c'tor in through `training_sequence1(std::move(init_field_sync_common<…>(false))` in the initializer list?

Reply:
Maybe. But at the cost of one indirection when the sequences are accessed.

I'd say there are three options:

1. As this PR currently does: Depending on compiler and options could do a needless copy, but the training data is inlined with the impl object so no indirection.
2. Switch to `std::vector`. No copies on init, but an indirection on access to the data, as `vector`s don't have a "small string optimization".
3. Call init functions, as before, but now with `std::array` instead of C arrays. No performance cost, as init is done in-place, and data inlined. Only downside is that it's a shame to not use the neatness of initializer list.

I've not played with dtv, but I suspect the important bit is to make the data path fast, which means (1) or (3). I'd lean towards (3).

Reply:
opted for (3)!